routes/review.js

```diff
@@ -0,0 +1,113 @@
+var express = require('express');
+var Reviews = require ('../database/reviews');
+var router = express.Router();
+
+router.route('/')
+
+	.get(function (req, res) {
+
+		Reviews.find({}, function (err, reviews) {
+			if (err) {
+				res.status(500).send({'error': err});
+			} else {
+				res.status(200);
+				var accept =req.get('Accept');
+				if(accept.indexOf("html")){
+					res.render('reviewsList', { title: 'Reviews', elements: reviews });
+				}
+				else{
+					res.send(reviews);
+				}
+			}
+		});
+	})
+
+	.post(function(req, res){
+		if(!req.body.name || !req.body.placeType || !req.body.stars){
+			res.status(400).send({ 'error' : 'Missing parameters'});
+		}
+		else{
+			var review = {
+				name: req.body.name,
+				placeType : req.body.placeType,
+				stars: req.body.stars,
+			};
+			Reviews.create(review, function(err, review){
+				if(err){
+					res.status(500).send({'error' : err});
+				}
+				else{
+					res.status(201);
+					var accept =req.get('Accept');
+					if(accept.indexOf("html")){
+						res.render('reviewSelected', { title: review.name, element: review });
+					}
+					else{
+						res.send(review);
+					}
+				}
+			});
+		}	
+
+	});
+
+router.route('/:index')
+	.get(function(req, res, next){
+		if (!req.params.index) {
+            res.status(400).send({'error': 'Invalid operation'});
+        } 
+        else{
+        	Reviews.findOne({_id: req.params.index}, function (err, review) {
+			if (err) {
+				res.status(500).send({'error': err});
+			} else {
+				if(!review){
+					res.status(404).send({'error' : 'Not found this review'})
+				}
+				else{
+					res.status(200);
+					var accept =req.get('Accept');
+					if(accept.indexOf("html")){
+						res.render('reviewSelected', { title: review.name, element: review });
+					}
+					else{
+						res.send(review);
+					}
+				}
+			}
+		});
+        }
+	})
+
+	.put(function(req, res){
+		if (!req.params.index) {
+            res.status(400).send({'error': 'Missing parameters'});
+        }
+        else{
+        	Reviews.findOneAndUpdate({_id: req.params.index}, req.body , function (err, reviews) {
+			if (err) {
+				res.status(404).send({'error': err});
```

Comment:
Ici err est truthy, donc on sait qu'un erreur vient de mongo, donc plutot qqch de plus technique, alors le status code doit etre un 500.

Suppossons qu'on teste l'url localhost:3000/reviews/54f4221269d89ea24afa2270 où 54f4221269d89ea24afa2270 est un ObjectId de Mongo (une chaine avec 15 characters). Si cette ObjectId n'est pas en base, il se passe rien coté mongo, il trouve pas d'objet et il fait pas d'update.

Si on teste avec localhost:3000/reviews/54f4221, on voit que 54f4221 n'a pas le format d'un ObjectId et l'erreur qu'on obtient est 

{
"error":{
"message": "Cast to ObjectId failed for value \"54f4221\" at path \"_id\"",
"name": "CastError",
"type": "ObjectId",
"value": "54f4221",
"path": "_id"
}
} 

Là, plutôt qu'un souci d’existence de ressource, il s'agit d'un erreur technique car on ne fournit pas un id avec la bonne syntaxe pour que la requête aie succès.